### PR TITLE
Rustdoc accessibility: use real headers for doc items

### PR DESCRIPTION
**Pull request by notriddle**
Part of #87059 

Partially reverts #84703

Preview at: https://notriddle.com/notriddle-rustdoc-test/real-headers/std/index.html

### CODE REVIEW

**Comment by ahicks92** [comment]
To get ahead of the "can you check it please": yes, this works at least with my setup, and it looks like you just made my day.

Also apparently I missed the table of contents headers one if that's not part of this one, but that's also very nice.

File: src/test/rustdoc/assoc-consts.rs [hunk 1]
```diff
@@ -13,7 +13,7 @@ pub trait Foo {
 pub struct Bar;
 
 impl Foo for Bar {
-    // @has assoc_consts/struct.Bar.html '//code' 'impl Foo for Bar'
+    // @has assoc_consts/struct.Bar.html '//h3' 'impl Foo for Bar'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/assoc-consts.rs:16]:
```suggestion
        // @has assoc_consts/struct.Bar.html '//h3[@class="code-header"]' 'impl Foo for Bar'
```

File: src/test/rustdoc/assoc-consts.rs [hunk 2]
```diff
@@ -77,7 +77,7 @@ pub trait Qux {
     const QUX_DEFAULT2: u32 = 3;
 }
 
-// @has assoc_consts/struct.Bar.html '//code' 'impl Qux for Bar'
+// @has assoc_consts/struct.Bar.html '//h3' 'impl Qux for Bar'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/assoc-consts.rs:80]:
```suggestion
// @has assoc_consts/struct.Bar.html '//h3[@class="code-header"]' 'impl Qux for Bar'
```

File: src/test/rustdoc/assoc-types.rs [hunk 3]
```diff
@@ -2,11 +2,11 @@
 
 // @has assoc_types/trait.Index.html
 pub trait Index<I: ?Sized> {
-    // @has - '//*[@id="associatedtype.Output"]//code' 'type Output: ?Sized'
+    // @has - '//*[@id="associatedtype.Output"]//h4' 'type Output: ?Sized'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/assoc-types.rs:5]:
```suggestion
        // @has - '//*[@id="associatedtype.Output"]//h4[@class="code-header"]' 'type Output: ?Sized'
```

File: src/test/rustdoc/assoc-types.rs [hunk 4]
```diff
@@ -2,11 +2,11 @@
...
     type Output: ?Sized;
-    // @has - '//*[@id="tymethod.index"]//code' \
+    // @has - '//*[@id="tymethod.index"]//h4' \
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/assoc-types.rs:7]: Same.

File: src/test/rustdoc/assoc-types.rs [hunk 5]
```diff
@@ -2,11 +2,11 @@
...
     //      "fn index<'a>(&'a self, index: I) -> &'a Self::Output"
-    // @has - '//*[@id="tymethod.index"]//code//a[@href="trait.Index.html#associatedtype.Output"]' \
+    // @has - '//*[@id="tymethod.index"]//h4//a[@href="trait.Index.html#associatedtype.Output"]' \
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/assoc-types.rs:9]: Same.

File: src/test/rustdoc/async-fn.rs [hunk 6]
```diff
@@ -35,9 +35,9 @@ pub async fn quux() -> impl Bar {
 }
 
 // @has async_fn/struct.Foo.html
-// @matches - '//code' 'pub async fn f\(\)$'
-// @matches - '//code' 'pub async unsafe fn g\(\)$'
-// @matches - '//code' 'pub async fn mut_self\(self, first: usize\)$'
+// @matches - '//h4' 'pub async fn f\(\)$'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/async-fn.rs:38]: Same.

File: src/test/rustdoc/async-fn.rs [hunk 7]
```diff
@@ -35,9 +35,9 @@ pub async fn quux() -> impl Bar {
...
+// @matches - '//h4' 'pub async unsafe fn g\(\)$'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/async-fn.rs:39]: Same.

File: src/test/rustdoc/async-fn.rs [hunk 8]
```diff
@@ -35,9 +35,9 @@ pub async fn quux() -> impl Bar {
...
+// @matches - '//h4' 'pub async fn mut_self\(self, first: usize\)$'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/async-fn.rs:40]: Same.

File: src/test/rustdoc/blanket-reexport-item.rs [hunk 9]
```diff
@@ -1,6 +1,6 @@
 #![crate_name = "foo"]
 
-// @has foo/struct.S.html '//div[@id="impl-Into%3CU%3E"]//code' 'impl<T, U> Into<U> for T'
+// @has foo/struct.S.html '//div[@id="impl-Into%3CU%3E"]//h3' 'impl<T, U> Into<U> for T'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/blanket-reexport-item.rs:3]: Same.

File: src/test/rustdoc/const-display.rs [hunk 10]
```diff
@@ -49,19 +49,19 @@ pub const unsafe fn bar_not_gated() -> u32 { 42 }
 pub struct Foo;
 
 impl Foo {
-    // @has 'foo/struct.Foo.html' '//div[@id="method.gated"]/code' 'pub fn gated() -> u32'
+    // @has 'foo/struct.Foo.html' '//div[@id="method.gated"]/h4' 'pub fn gated() -> u32'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/const-display.rs:52]: Same.

**Comment by GuillaumeGomez** [comment]
The notable tooltip is broken too:

Screenshot from 2021-07-23 11-52-17

**Comment by notriddle** [comment]
@GuillaumeGomez okay, I fixed the tooltip and all of the test cases.

File: src/test/rustdoc/const-generics/const-generics-docs.rs [hunk 11]
```diff
@@ -19,10 +19,10 @@ pub use extern_crate::WTrait;
 
 // @has foo/trait.Trait.html '//pre[@class="rust trait"]' \
 //      'pub trait Trait<const N: usize>'
-// @has - '//*[@id="impl-Trait%3C1_usize%3E-for-u8"]//code' 'impl Trait<1_usize> for u8'
-// @has - '//*[@id="impl-Trait%3C2_usize%3E-for-u8"]//code' 'impl Trait<2_usize> for u8'
-// @has - '//*[@id="impl-Trait%3C{1%20+%202}%3E-for-u8"]//code' 'impl Trait<{1 + 2}> for u8'
-// @has - '//*[@id="impl-Trait%3CN%3E-for-%5Bu8%3B%20N%5D"]//code' \
+// @has - '//*[@id="impl-Trait%3C1_usize%3E-for-u8"]//h3' 'impl Trait<1_usize> for u8'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/const-generics/const-generics-docs.rs:22]: I also meant for all other h3/h4. :)

Comment (notriddle) [reply src/test/rustdoc/const-generics/const-generics-docs.rs]: Okay, @GuillaumeGomez, that's done.

**Comment by GuillaumeGomez** [comment]
Funnily enough, you fixed the background color issue when an item is active:

Screenshot from 2021-07-24 21-12-23

However, it creates an issue when you hover it:

Screenshot from 2021-07-24 21-12-25

Not a blocker though, let's send another fix once this PR has been merged.

Thanks a lot! r=me once CI pass.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
That seems a like a "real" error. From the diff, I can't see where you forgot to close a tag though... :-/

**Comment by notriddle** [comment]
BTW, how do I run that locally? I couldn’t find any information on it anywhere, and most of the obvious ones like `./x.py doc` didn’t do it.

**Comment by GuillaumeGomez** [comment]
`x.py test src/tools/html-checker`. :)

**Comment by notriddle** [comment]
Protip: do not nest header inside of other headers. Tidy hates that, and it gives very confusing error messages when you do it.

**Comment by pietroalbini** [comment]
This has been merged.

**Comment by pietroalbini** [comment]
Actually, this will also need to be backported to 1.55.0 beta as it was merged after the beta cutoff.

**Comment by pietroalbini** [comment]
Included in #87508

**Pull request closed without merging**